permissions/src/main/java/com/google/accompanist/permissions/MutablePermissionState.kt

```diff
@@ -47,112 +44,66 @@ import kotlinx.coroutines.channels.Channel
+    val currentActivity by rememberUpdatedState(context.findActivity())
 
-    fun checkPermission(): Boolean =
-        ContextCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED
-
-    val permissionState = remember { mutableStateOf(checkPermission()) }
+    // FIXME: This should probably be rememberSaveable with a saver for the
+    //  permissionRequested value
+    val state = remember(permission) { MutablePermissionState(permission = permission) }
 
     // Check if the permission was granted when the app comes from the background
     // The user might've gone to the Settings screen and granted the permission
     // We don't check if the permission was denied as that triggers a process restart
-    val lifecycle = LocalLifecycleOwner.current.lifecycle
     val permissionCheckerObserver = remember(permission) {
         LifecycleEventObserver { _, event ->
             if (event == Lifecycle.Event.ON_START) {
-                if (!permissionState.value) { permissionState.value = checkPermission() }
+                if (!state.hasPermission) {
+                    state.hasPermission = context.checkPermission(permission)
+                }
             }
         }
     }
+    val lifecycle = LocalLifecycleOwner.current.lifecycle
     DisposableEffect(lifecycle, permissionCheckerObserver) {
         lifecycle.addObserver(permissionCheckerObserver)
-        onDispose {
-            lifecycle.removeObserver(permissionCheckerObserver)
+        onDispose { lifecycle.removeObserver(permissionCheckerObserver) }
+    }
+
+    // Observe the state refresh channel, to trigger a refresh when requested.
+    LaunchedEffect(state) {
+        for (refreshEvent in state.refreshChannel) {
+            state.shouldShowRationale = currentActivity.shouldShowRationale(permission)
         }
     }
 
-    return permissionState
+    return state
 }
 
 /**
- * Creates a [ShouldShowRationaleState] that is remembered across recompositions.
+ * A mutable state object that can be used to control and observe permission status changes.
+ *
+ * In most cases, this will be created via [rememberMutablePermissionState].
+ *
+ * @param permission the permission to control and observe.
  */
-@Composable
-private fun rememberShouldShowRationaleState(permission: String): ShouldShowRationaleState {
-    val currentActivity by rememberUpdatedState(LocalContext.current.findActivity())
+@Stable
+internal class MutablePermissionState(
+    override val permission: String,
+) : PermissionState {
+    override var hasPermission: Boolean by mutableStateOf(false)
+    override var shouldShowRationale: Boolean by mutableStateOf(false)
+    override var permissionRequested: Boolean by mutableStateOf(false)
 
-    fun shouldShow(): Boolean =
-        ActivityCompat.shouldShowRequestPermissionRationale(currentActivity, permission)
+    // This feels a bit memory leaky
+    internal lateinit var launcher: ActivityResultLauncher<String>
```

Comment:
It looks indeed a bit memory leaky. I think we can solve this by not having the `rememberMutablePermissionState` and have that logic inside `rememberPermissionState`. I don't know how that'd influence the multiple requests flow. 

I'll experiment with this but looks good so far! Thank you

Reply:
Cool, I'll let you merge this into your branch. I'll close this PR.